### PR TITLE
[DEV-1168] manage webinars subscription withuser preferences

### PR DESCRIPTION
**Pull request by MarcoPonchia**
#### List of Changes

Connect webinar button to user preferences

#### Motivation and Context

Connect webinar button to user preferences

#### How Has This Been Tested?

User can test this PR in homepage: http://localhost:3000/ or webinar detail page: http://localhost:3000/webinars/comunicazioni-a-valore-legale
To check with multiple webinar replace file `apps/nextjs-website/src/_contents/webinars.ts` with this example:
```
import { Webinar } from '@/lib/types/webinar';
import { appIOGuideListsPath } from './appIo/guideListsPath';
import { appIoQuickStartGuidePath } from './appIo/quickStartGuidePath';

export const webinars: readonly Webinar[] = [
  {
    title: 'Esplorando App IO: Le nuove API dei Servizi',
    description:
      'Il mondo di IO è molto ampio, il primo passo è capire come integrarsi alle nuove API e con questo webinar vedremo le informazioni principali per farlo!',
    html:
      `<h4 style="font-weight: 600; font-size: 24px;">Tutto quello che serve per integrarsi con le nuove API</h4> \n` +
      `<p>Durante il webinar, approfondiremo la creazione, l'invio di un servizio e la gestione di scenari comuni, come "Ho creato un servizio e lo vedo in stato di bozza. Cosa devo fare?". Partecipate a questo webinar per ottenere una comprensione approfondita e per imparare a gestire in modo efficace il ciclo di vita dei vostri servizi su IO. </p> \n` +
      `<br /> \n` +
      `<img src='/images/webinar-image.png' /> \n` +
      `<br /> \n` +
      `<p>Gli interventi dei nostri esperti saranno intervallati da momenti di Q&A per condividere le migliori pratiche per un'esperienza fluida e di successo nella gestione dei vostri servizi. </p> \n`,
    slug: 'comunicazioni-a-valore-legale',
    speakers: [
      {
        name: 'Ivan Diana',
        jobTitle: 'Engineering Manager',
        description:
          'Engineering leader appassionato, metto al primo posto le persone e poi la tecnologia',
        imagePath: '/images/speaker-diana.png',
      },
      {
        name: 'Lorenzo Franceschini',
        jobTitle: 'Software Engineer - Service Management App IO',
        description:
          'Sviluppatore con esperienza decennale, formatore per passione. Mi piace il confronto costruttivo e la possibilità di insegnare e apprendere dagli altri',
        imagePath: '/images/speaker-franceschini.png',
      },
      {
        name: 'Demetrio Ferraro',
        jobTitle: 'Operations Specialist - Service Management App IO ',
        description:
          'Customer specialist execution ho fatto della trasversalità e complementarità il mio punto forte',
        imagePath: '/images/speaker-ferraro.png',
      },
    ],
    startDateTime: '2023-12-14T11:00:00',
    endDateTime: '2023-12-14T11:40:00',
    subscribeCtaLabel: '',
    startInfo: {
      title: 'Risorse correlate',
      cards: [
        {
          title: 'Quick Start',
          text: 'Aderire a IO tramite l’Area Riservata, creare un servizio, verificare l’esistenza di un utente, inviare un messaggio: ecco come si fa',
          href: `${appIoQuickStartGuidePath.path}`,
          iconName: 'FlagOutlined',
        },
        {
          title: 'Documentazione API',
          text: "Esplora le API Rest per l'invio dei messaggi e la creazione di servizi sull'app IO",
          href: '/app-io/api',
          iconName: 'FolderOutlined',
        },
        {
          title: 'Guida tecnica',
          text: 'Cosa può fare il tuo ente con l’app IO? Seleziona il settore di appartenenza e approfondisci i principali casi d’uso dedicati.',
          href: `${appIOGuideListsPath.path}/io-guida-tecnica/v4.0`,
          iconName: 'MenuBook',
        },
      ],
    },
    relatedLinks: [
      {
        path: `${appIOGuideListsPath.path}/supporto-agli-enti/v1.0`,
        name: 'Consulta FAQ e approfondimenti nella documentazione di supporto agli Enti',
      },
      {
        path: 'https://files.gitbook.com/v0/b/gitbook-x-prod.appspot.com/o/spaces%2FZugEhXfhuchEaSiqU3p5%2Fuploads%2FKkqRR4GLnQtJ04He1afT%2FAccordo%20Base%20di%20Adesione%20App%20IO_Standard_Ver.2.4.16_1_2023.docx.pdf?alt=media',
        name: 'Scarica il contratto di adesione a IO',
      },
      {
        path: `${appIOGuideListsPath.path}/kit-comunicazione/v1.0`,
        name: 'Leggi kit di comunicazione',
      },
    ],
  },
  {
    title: 'web-2',
    description:
      'Raccogli tutti i servizi digitali del tuo ente in un’unica piattaforma e interagisci in modo semplice e sicuro con i cittadini. Raccogli tutti i servizi digitali del tuo ente in un’unica piattaforma.',
    html:
      `<h4 style="font-weight: 600; font-size: 24px;">Sottotitolo</h4> \n` +
      `<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p> \n` +
      `<br /> \n` +
      `<img src='/images/webinar-image.png' /> \n` +
      `<br /> \n` +
      `<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p> \n`,
    slug: 'web-2',
    subscribeCtaLabel: 'Test',
    speakers: [
      {
        name: 'Anna Bianchi',
        jobTitle: 'Product Owner SEND',
        description:
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam',
        imagePath: '/images/speaker-00.png',
      },
    ],
    startDateTime: '2024-12-31T11:00:00',
    endDateTime: '2024-12-31T13:00:00',
  },
  {
    title: 'web-3 molto lungo',
    description:
      'Raccogli tutti i servizi digitali del tuo ente in un’unica piattaforma e interagisci in modo semplice e sicuro con i cittadini. Raccogli tutti i servizi digitali del tuo ente in un’unica piattaforma.',
    html:
      `<h4 style="font-weight: 600; font-size: 24px;">Sottotitolo</h4> \n` +
      `<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p> \n` +
      `<br /> \n` +
      `<img src='/images/webinar-image.png' /> \n` +
      `<br /> \n` +
      `<p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum</p> \n`,
    slug: 'KmH1sF0rD5UwyXMDMbIUPq4KDSWpmMf5PUOVxWQ6RvpxePPsUYZM8hpffRHZiZ5ihp12g0kj0xDVPA3fdGJyszBfEID8eQZCku1txneJwp3q7wzlxy2WVUptrRYTs1yF3TLyoeh1IOLnblUJAOv4t2s1xKyN04eiSagdHZriOrhKsDyPyrM156Jbc1xI8Ww73RRApDPikclDcj1eweItdeaWyjjRqO4MSQYc3oMGbK9WygRLjZkaMEcl918s1O9WUhjxlWAtQVZ2Y7YcPyMPrvNfvpvZKTwYYrDKwkWIDmMlecQonrXeU7N5L4JNwYXnYjy8lSCwspp0AcZlyoKqdGNsuqpBKnkzrasBQUGdCNGpGmHFFTGzO5W7LKyfIesyoXmazJMuZ5up6BNLN2zNIzfN9ZWpKoGhclrvpmI4Bz9BdjBu7jIJcLytQq3kLz8IMTwOF1pWCCq1KxIWNmVfqdan1udfgdVDdtDVmDihz6m3HIPvjBFldqh1rpgQIDPM2zYBIBfLiTBp2MDVHJ7LjJkXwr076z8gbIpKy5BiqPB8qsLDjODzUZSBwCB1XtIbIlR1EYihnQkfeEc5RjAJZ5X8Timry9gOdg8hsiaHUTVrZ14BNYOyepIrxfkGdRketJT5xCHaplzxPfbSvCNeGpjE3sMD38sFZGoyuZJdY8sMxTl2lHbMdP3xmUQf1Qy1frP7C9J2k4h2MOpiSivNC4msc8Fzl4nZHiwR83OBJ8Rc9RmfOL01IvchPUqzQ5MxOrdiYkBoITXLocadmlK884C6Btf7oaZOyKQDi0cmRKQFuPyK9FJSYdj6gptieLtlVj01qPBVcK5UJIeuJr4xXxHg9slHozlqSWb99P5uXckm2UHdaKl3CXWzkfvqczjieCq0lolEVccfJHWiyZhDFdN0h6y8fDW0GKL6kye2vuIpvzdX3loIGSSrLW3iD1fkHgcTVRl7Gjz7np4cCczoAErspbsXU1yTsxdESwcyHxOe4KGHZTQGS6n5TtVIoc64bQc0mfsknXkfE1wpUVGPvlOXGHDOGlTyGQFA0P8t9b2moL1IMb62rMiWWC9ZRnTZyW1ayCf73dFyt6Fx5mWxRCTX0OkCF42tDIB9gLLTKzGxxVyne2P064EdTRv5qL04RNMlWAqLPdH6VNYjTCBuxudriIV9AcZcioZ9UDEBtSqGlLwIZBpPn0o514F9Log0S3fMhuY2uMJUss6gzhRcWHujlpinUxnoZiMzZKLrAIYHepuo8OsXnQ4HgbPWZnXbVcV2bn5sgSeqxJYL1sKkn4zC7SYvIVKkeMLEypbpKjGpaTaFRefQnnGp47i9acs8kn5L0H0aFQmoKleLkVp1pNkjOjLIZ2wDTQKzE4MvzgtMEz26sv9n8KyqjOtrFdQXcL6bIHqTIr002MDG74jODDgwrKqTXrjdf9ScRufWL8jXvg3mQGBTFDMqplcY1bUwFcFU1KHsSOBloasEmHYk13lJE2u1WARR2Fpy7bMHPO3xtptDSg7TXwNcWFaw5bzMI8YeLuADCn2WuVWDqoW0KJeGHq4TPcD0mymY4mgqsQd3F0FRGWyc5qk9Yq0LH5w2cUobikuKcuIBJGmYUeStD9RXPbbokXcq8Z53BCb5DiC4PphuswVIT9x8zXFwxZupHmh2o2XgcMFzB5D40We7u6qI8y9BPpOWchWvf4ZYvYZnF15vdGiowIBDlb6PMavoyMBe322yGg7ZdCJCqal2jiJ89N844PpthbVmg4zNQO0qi5MfX0xKASmieao4jQ534zkj0zxNROq5yRXzV76VZdPb5wCSYYWxgvZhFI5Lhpotr7drrp82BJbF6pSYI8KM2X2TAVB0k7n2xWG4jlN0GO3BdJCKr8O8oOJKGB53TKl52sG7c3NdihnJC2sxGuL1FhdscLVrNYGmx8jaC4xw5ELcq0bDT6T1YDcktyOjppvYNVKu19Mbtr0TCVXy9bq7i2mTSJVBDg4Lckgv6Z5hogg734WTKXU3B9xRe8yOrwiz17QAGVHHAI6yWxk3D27X',
    speakers: [
      {
        name: 'Anna Bianchi',
        jobTitle: 'Product Owner SEND',
        description:
          'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam',
        imagePath: '/images/speaker-00.png',
      },
    ],
    startDateTime: '2024-12-31T11:00:00',
    endDateTime: '2024-12-31T13:00:00',
  },
];

```

#### Screenshots (if appropriate):
[Screencast from 23-11-2023 15:56:42.webm](https://github.com/pagopa/developer-portal/assets/24475319/1ff08c58-51a7-41ca-9d44-1a069ecb1d8d)
[screencast-localhost_3000-2023.11.23-15_55_58.webm](https://github.com/pagopa/developer-portal/assets/24475319/de1809f1-19ba-44e4-8d36-27576ff84fc8)

#### Types of changes

- [x] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: apps/nextjs-website/src/__tests__/helpers/userPreferences.helpers.test.ts [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+  ).toEqual(newAttributes);
+});
+
+it('should return true if webinar subscription exists', () => {
+  expect(webinarSubscriptionPresent(existingSlugs[0], attributes)).toEqual(
+    true
+  );
+});
+
+it('should return false if webinar subscription not exists', () => {
```
Comment (marcobottaro) [inline apps/nextjs-website/src/__tests__/helpers/userPreferences.helpers.test.ts:153]:
```suggestion
it('should return false if webinar subscription does not exist', () => {
```

File: apps/nextjs-website/src/helpers/userPreferences.helpers.ts [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+export const MAX_USER_PREFERENCES_LENGTH = 2048;
+
+const DEFAULT_USER_PREFERENCES: UserPreferences = {
+  subscribedWebinarSlugs: [],
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function isUserPreferences(obj: any): obj is UserPreferences {
+  return obj?.subscribedWebinarSlugs !== undefined;
+}
```
Comment (marcobottaro) [inline apps/nextjs-website/src/helpers/userPreferences.helpers.ts:12]: :+1:

File: apps/nextjs-website/src/helpers/userPreferences.helpers.ts [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        subscribedWebinarSlugs: [
+          ...currentUserPreferences.subscribedWebinarSlugs,
+          slug,
+        ],
+      },
+      attributes
+    );
+  } else {
+    return attributes;
+  }
```
Comment (marcobottaro) [inline apps/nextjs-website/src/helpers/userPreferences.helpers.ts:69]:
```suggestion
  if (currentUserPreferences.subscribedWebinarSlugs.includes(slug)) {
    return attributes;
  }
  
  return mergeUserPreferencesToAttributes(
    {
      subscribedWebinarSlugs: [
        ...currentUserPreferences.subscribedWebinarSlugs,
        slug,
      ],
    },
    attributes
  );
```

File: apps/nextjs-website/src/helpers/userPreferences.helpers.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        subscribedWebinarSlugs:
+          currentUserPreferences.subscribedWebinarSlugs.filter(
+            (subscriptionSlug) => subscriptionSlug !== slug
+          ),
+      },
+      attributes
+    );
+  } else {
+    return attributes;
+  }
```
Comment (marcobottaro) [inline apps/nextjs-website/src/helpers/userPreferences.helpers.ts:89]:
```suggestion
  if (!currentUserPreferences.subscribedWebinarSlugs.includes(slug)) {
    return attributes;
  }
  
return mergeUserPreferencesToAttributes(
  {
    subscribedWebinarSlugs:
      currentUserPreferences.subscribedWebinarSlugs.filter(
        (subscriptionSlug) => subscriptionSlug !== slug
      ),
  },
  attributes
);
```

File: apps/nextjs-website/src/helpers/userPreferences.helpers.ts [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+export function webinarSubscriptionPresent(
```
Comment (marcobottaro) [inline apps/nextjs-website/src/helpers/userPreferences.helpers.ts:92]:
```suggestion
export function webinarSubscriptionExists(
```

**Review by marcobottaro — CHANGES_REQUESTED** [review]

**Review by datalek — CHANGES_REQUESTED** [review]
Once subscribed to the webinar, the button in the homepage is not updated once the page is refreshed

**Comment by MarcoPonchia** [comment]
> Once subscribed to the webinar, the button in the homepage is not updated once the page is refreshed

Fixed 
[Screencast from 24-11-2023 10:38:46.webm](https://github.com/pagopa/developer-portal/assets/24475319/f6ed8772-8acb-49d1-9d43-903c77034259)

**Review by kin0992 — APPROVED** [review]

**Pull request merged**